src/content/en/fundamentals/engage-and-retain/push-notifications/sending-messages.markdown

```diff
@@ -3,93 +3,343 @@ layout: shared/narrow
+twenty-four hours. This can be calculated by converting the current date to 
+milliseconds and adding the duration. For example, in Node.js you could do this: 
+   
+{% highlight javascript %}
+Math.floor((Date.now() / 1000) + 12 * 60 * 60)
+{% endhighlight %}
+
+This should match the value in the TTL header.
+
+**sub**  
+Specifies a subject, which the VAPID spec defines as a way for the push service 
+to contact a message sender. This can be a URL or a mail to URL.  
+  
+A complete payload looks like this:
+
+{% highlight javascript %}  
+{  
+  "aud": "http://push.example.net",  
+  "exp": "1469618703",  
+  "sub": "mailto: my-email@some-url.com"  
+}
+{% endhighlight %}
+
+#### Signature
+
+Create the signature by first concatenating the JWT header and the payload with 
+a dot, the encrypting the private key you created earlier. We're not going to 
+show you how to do this, but there are a number of encryption libraries 
+available. The result will look something like the following:  
+  
+<pre>Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJhdWQiOiJodHRwczovL2ZjbS5nb29nbGVhcGlzLmNvbSIsImV4cCI6MTQ2NjY2ODU5NCwic3ViIjoibWFpbHRvOnNpbXBsZS1wdXNoLWRlbW9AZ2F1bnRmYWNlLmNvLnVrIn0.Ec0VR8dtf5qb8Fb5Wk91br-evfho9sZT6jBRuQwxVMFyK5S8bhOjk8kuxvilLqTBmDXJM5l3uVrVOQirSsjq0A</pre>
+
+### Message payload
+
+When implementing server code there are two methods of sending messages to 
+consider.  
 
 * Messages with a data payload.
 * Messages without a data payload, often called a tickle.
 
-In the case of the tickle, the service worker uses the message as a signal to
-fetch data from an endpoint. The [Handling Messages](handling-messages) section
-contains sample code for showing how a service worker does this.
-
-Why would you ever send a message without a payload? There are two reasons.
+In the case of the tickle, the service worker uses the message as a signal to 
+fetch data from an endpoint. The Handling Messages section contains sample code 
+for showing how a service worker does this.  
+  
+Why would you ever send a message without a payload? There are two reasons.  
 
 * You need to send something larger than the 4k payload limit set by the spec
 * The clients needs fresher data than what would be in the push.
 
-Technically, there's another reason which is that browser capabilities may vary
-for a while yet, but the two main reasons will likely always apply. If the
+Technically, there's another reason which is that browser capabilities may vary 
+for a while yet, but the two main reasons will likely always apply. If the 
 browser doesn't support payloads, the subscription object won't contain keys.
 
-## Sending messages {#sending-messages}
+The payload, regardless of how you get it to the client, should be encrypted. 
+For payloads sent through the message server you must encrypt it using the 
+publicKey and the authentication secret. It must also be padded, and salted with 
+random 16 bytes, urlBase64 encoded. Finally, it's added to the body of the 
```

Comment:
this doesn't read well. "and salted with 16 random bytes and then url safe base64 encoded."